TombEngine/Game/effects/drip.cpp

```diff
@@ -1,130 +1,173 @@
 
-			drip.pos += drip.velocity;
-			float normalizedAge = drip.age / drip.life;
-			drip.color = Vector4::Lerp(DRIP_COLOR, Vector4::Zero, normalizedAge);
-			drip.height = Lerp(DRIP_WIDTH / 0.15625f, 0, normalizedAge);
-			short room = drip.room;
-			FloorInfo* floor = GetFloor(drip.pos.x, drip.pos.y, drip.pos.z, &room);
-			int floorheight = floor->FloorHeight(drip.pos.x, drip.pos.z);
-			int waterHeight = GetWaterHeight(drip.pos.x, drip.pos.y, drip.pos.z, drip.room);
+			float verticalVel = isSmallSplash ?
+				Random::GenerateFloat(VERTICAL_VELOCITY_LOW_MIN, VERTICAL_VELOCITY_LOW_MAX) :
+				Random::GenerateFloat(VERTICAL_VELOCITY_HIGH_MIN, VERTICAL_VELOCITY_HIGH_MAX);
+			auto vel = (direction * VELOCITY_BASE) + Vector3(0.0f, -verticalVel, 0.0f);
 
-			if (drip.pos.y > floorheight) 
-				drip.active = false;
+			float gravity = isSmallSplash ?
+				Random::GenerateFloat(GRAVITY_LOW_MIN, GRAVITY_LOW_MAX) :
+				Random::GenerateFloat(GRAVITY_HIGH_MIN, GRAVITY_HIGH_MAX);
 
-			if (drip.pos.y > waterHeight) 
-			{
-				drip.active = false;
-				SetupRipple(drip.pos.x, waterHeight, drip.pos.z, Random::GenerateInt(16, 24), RIPPLE_FLAG_SHORT_INIT | RIPPLE_FLAG_LOW_OPACITY);
-			}
+			SpawnDrip(dripPos, roomNumber, vel, LIFE_MAX, gravity);
 		}
 	}
 
-	DripParticle& getFreeDrip()
+	void SpawnWetnessDrip(const Vector3& pos, int roomNumber)
 	{
-		for (int i = 0; i < dripParticles.size(); i++)
-		{
-			if (!dripParticles[i].active)
-				return dripParticles[i];
-		}
-
-		return dripParticles[0];
+		constexpr auto LIFE_MAX		= 1.0f;
+		constexpr auto GRAVITY_MAX	= 6.0f;
+		constexpr auto GRAVITY_MIN	= 3.0f;
+		constexpr auto SPAWN_RADIUS = BLOCK(1 / 32.0f);
+
+		auto sphere = BoundingSphere(pos, SPAWN_RADIUS);
+		auto dripPos = Random::GeneratePointInSphere(sphere);
+		SpawnDrip(dripPos, roomNumber, Vector3::Zero, LIFE_MAX, Random::GenerateFloat(GRAVITY_MIN, GRAVITY_MAX));
 	}
 
-	void SpawnWetnessDrip(const Vector3& pos, int room)
+	void UpdateDrips()
 	{
-		auto& drip = getFreeDrip();
-		drip = {};
-		drip.active = true;
-		drip.pos = pos;
-		drip.room = room;
-		drip.life = DRIP_LIFE;
-		drip.gravity = Random::GenerateFloat(3, 6);
-	}
+		constexpr auto RIPPLE_SIZE_WATER_MAX  = 24.0f;
+		constexpr auto RIPPLE_SIZE_WATER_MIN  = 16.0f;
+		constexpr auto RIPPLE_SIZE_GROUND_MAX = 16.0f;
+		constexpr auto RIPPLE_SIZE_GROUND_MIN = 8.0f;
+		constexpr auto RIPPLE_HEIGHT_OFFSET	  = 4;
 
-	void SpawnSplashDrips(const Vector3& pos, int number, int room)
-	{
-		for (int i = 0; i < number; i++) 
+		if (Drips.empty())
+			return;
+
+		for (auto& drip : Drips)
 		{
-			auto dripPos = pos + Vector3(Random::GenerateFloat(-128, 128), Random::GenerateFloat(-128, 128), Random::GenerateFloat(-128, 128));
-			auto direction = (dripPos - pos);
-			direction.Normalize();
+			if (drip.Life <= 0.0f)
+				continue;
+
+			// Update velocity.
+			drip.Velocity.y += drip.Gravity;
+			if (TestEnvironment(ENV_FLAG_WIND, drip.RoomNumber))
+				drip.Velocity += Weather.Wind();
```

Comment:
this line caused crash for Adngel

Reply:
Odd, a similar block was used in the old drip implementation. Will look into it.